### ISSUE TITLE
"False positive" on duplicate array key when destructing array into array

### ISSUE THREAD

**Issue by 7ochem** [vimeo/psalm#8616]
The code snippet shows two cases. It is giving two "DuplicateArrayKey - Key 'b' already exists on array" errors. In the second case it is certainly really helpful, because it actually detects a flaw in the code. In the **first case**, I think the code is completely legit and the DuplicateArrayKey error reported is not helping to improve the code and (in my opinion) should not be reported.

https://psalm.dev/r/9e3858d4cc

**Comment by orklah** [comment]
> the code is completely legit 

There certainly is cases where it can be useful to do that. But is it more likely to report legit cases than errors by the developer? It's debatable. 

Creating an edge case here would also cascade into a lot others. array_merge is likely concerned too. What if the duplicate coming from the array destructuring is actually positionned after the "hard coded" one? etc...

**Comment by 7ochem** [comment]
Ok, so maybe it's just better to add a `/** @psalm-suppress DuplicateArrayKey */` when it is intentional?

**Comment by orklah** [comment]
Yeah, that would be the best, you could also disable this issue for a whole file if needed through the issueHandler in your config. I'll close this for now

**Comment by kylekatarnls** [comment]
Can we take another look at that?

`[...$a, ...$b]` is just a shortcut for `array_merge($a, $b)` however Psalm does not complain when using `array_merge`:

https://psalm.dev/r/b046c26ca9

So I suspect this error does not happen by design but by accident.

In the example of @7ochem `'b'` key is "lost" since `$array1` is overridden but not that `DuplicateArrayKey` will also happen with:
```php
$defaults = ['a' => 'default value', 'b' => 'default value'];

$choices = [...$defaults, 'b' => 'my choice'];

// I will later use $choices
// I will later also use $default
```

I approve the initial message:
`DuplicateArrayKey` in the `$array2` of [the first snippet](https://psalm.dev/r/9e3858d4cc) is always dead code.

But in the code above with `$defaults` and `$choices`, `'b'` is just not duplicated, it's overridden and both values will be used.

**Comment by distantnative** [comment]
@orklah Could we please revisit this? It really doesn't make much sense that the use case is completely acceptable when using `array_merge` but when using the equivalent syntax, Psalm starts throwing false errors all over the place. I'd really appreciate if this would be reconsidered. Thanks for your work!

**Comment by weirdan** [comment]
There's at least two idiomatic ways to express it Psalm doesn't complain about: https://psalm.dev/r/3a76a35084

**Comment by BafS** [comment]
The issue is still there https://psalm.dev/r/8fc8d87f2b, both are valid

**Comment by 7ochem** [comment]
> There's at least two idiomatic ways to express it Psalm doesn't complain about: https://psalm.dev/r/3a76a35084

That actually supports the point I was making: those options are the same idea (reshaping an array) in a different spelling. If Psalm is fine with them but flags the "destructuring-into-array" form as `DuplicateArrayKey`, the difference is about how Psalm handles that syntax, not about a clearly wrong pattern. I’d still like Psalm to accept the version I used, same as these idioms.

@danog you have an opinion on this?
